source/distributed/tria.cc

```diff
@@ -3733,39 +3633,48 @@ namespace parallel
       std::vector<unsigned int> weights;
       weights.reserve(this->n_active_cells());
 
-      // Recurse over p4est and Triangulation
+      // Iterate over p4est and Triangulation relations
       // to find refined/coarsened/kept
       // cells. Then append cell_weight.
       // Note that we need to follow the p4est ordering
       // instead of the deal.II ordering to get the cell_weights
       // in the same order p4est will encounter them during repartitioning.
-      for (unsigned int c=0; c<this->n_cells(0); ++c)
+      for (const auto &it_rel : local_quadrant_cell_relations)
         {
-          // skip coarse cells, that are not ours
-          if (tree_exists_locally<dim,spacedim>(parallel_forest,c) == false)
-            continue;
+          const auto &cell_status = std::get<1>(it_rel);
+          const auto &cell_it = std::get<2>(it_rel);
+
+          switch (cell_status)
+            {
+            case parallel::distributed::Triangulation<dim,spacedim>::CELL_PERSIST:
+              weights.push_back(1000);
+              weights.back() += this->signals.cell_weight(cell_it,
+                                                          parallel::distributed::Triangulation<dim,spacedim>::CELL_PERSIST);
+              break;
 
-          const unsigned int coarse_cell_index =
-            p4est_tree_to_coarse_cell_permutation[c];
+            case parallel::distributed::Triangulation<dim,spacedim>::CELL_REFINE:
+            case parallel::distributed::Triangulation<dim,spacedim>::CELL_INVALID:
+            {
+              // calculate weight of parent cell
+              unsigned int parent_weight(1000);
+              parent_weight += this->signals.cell_weight(cell_it,
+                                                         parallel::distributed::Triangulation<dim,spacedim>::CELL_REFINE);
+              // assign the weight of the parent cell equally to all children
+              weights.push_back(parent_weight);
+              break;
+            }
 
-          const typename Triangulation<dim,spacedim>::cell_iterator
-          dealii_coarse_cell (this, 0, coarse_cell_index);
+            case parallel::distributed::Triangulation<dim,spacedim>::CELL_COARSEN:
+              weights.push_back(1000);
+              weights.back() += this->signals.cell_weight(cell_it,
+                                                          parallel::distributed::Triangulation<dim,spacedim>::CELL_COARSEN);
+              break;
+
+            default:
+              Assert(0, ExcInternalError());
+              break;
+            }
 
-          typename dealii::internal::p4est::types<dim>::quadrant p4est_coarse_cell;
-          dealii::internal::p4est::functions<dim>::
-          quadrant_set_morton (&p4est_coarse_cell,
-                               /*level=*/0,
-                               /*index=*/0);
-          p4est_coarse_cell.p.which_tree = c;
```

Comment:
I did not set the quadrant Morton indices in the <code>compute_quadrant_cell_relations</code> function. Instead, I used <code>init_coarse_quadrant</code>, which seems to fulfil a similar task, since all tests pass. So I didn't bother keeping it.